src/Orchard.Web/Core/Settings/Descriptor/ShellDescriptorManager.cs

```diff
@@ -50,47 +62,84 @@ public class ShellDescriptorManager : Component, IShellDescriptorManager {
             return descriptor;
         }
 
+        private const string EvictSignalName =
+            "ShellDescriptorRecord_EvictCache";
+        private const string DescriptorCacheName =
+            "Orchard.Core.Settings.Descriptor.ShellDescriptorManager.ShellDescriptorRecord";
         private ShellDescriptorRecord GetDescriptorRecord() {
-            return _shellDescriptorRepository.Get(x => x != null);
+            // fetching the ShellDescriptorRecord also causes NHibernate to launch
+            // SELECT queries to fetch the ShellFeatureRecords and the ShellParameterRecords.
+            // If we cache all that, we save those three select queries on every request.
+            // We should be careful in the eviction policy when the ShellDescriptorRecord
+            // gets updated.
+            return _cacheManager.Get(DescriptorCacheName, true, ctx => {
+                ctx.Monitor(_signals.When(EvictSignalName));
+                return _shellDescriptorRepository.Get(x => x != null);
+            });
         }
 
-        public void UpdateShellDescriptor(int priorSerialNumber, IEnumerable<ShellFeature> enabledFeatures, IEnumerable<ShellParameter> parameters) {
-            ShellDescriptorRecord shellDescriptorRecord = GetDescriptorRecord();
-            var serialNumber = shellDescriptorRecord == null ? 0 : shellDescriptorRecord.SerialNumber;
-            if (priorSerialNumber != serialNumber)
-                throw new InvalidOperationException(T("Invalid serial number for shell descriptor").ToString());
+        private string _lockString = "";
+        private string LockString {
+            get {
+                if (string.IsNullOrWhiteSpace(_lockString)) {
```

Comment:
Can be set once in the constructor. Can also be simplified (just tenant name + "ShellDescriptorManager").